zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchRepository.java

```diff
@@ -195,6 +193,45 @@ public CompletableFuture<Void> reindexDocuments(
         .thenApplyAsync(ignored -> null, executor);
   }
 
+  private List<String> fetchIndexMatchingIndexes(final String indexWildCard) throws IOException {
+    final List<String> indices = new ArrayList<>();
+    client
+        .cat()
+        .indices()
+        .thenAccept(
+            response ->
+                response.valueBody().stream()
+                    .map(IndicesRecord::index)
+                    .filter(index -> index.matches(indexWildCard))
+                    .forEach(indices::add))
+        .join();
```

Comment:
:x: Joining here will pin the background thread indefinitely. This applies as well to the ES repository version, which also joins the result of fetching the indices.

Instead, we should compose the operations:

```java
  @Override
  public CompletableFuture<Void> setLifeCycleToAllIndexes() {
    if (!retention.isEnabled()) {
      return CompletableFuture.completedFuture(null);
    }

    final var indexWildCard = "^" + indexPrefix + INDEX_WILDCARD;
    return fetchIndexMatchingIndexes(indexWildCard)
      .thenComposeAsync(indices -> setIndexLifeCycle(indices.toArray(String[]::new)), executor);
  }

  private CompletableFuture<List<String>> fetchIndexMatchingIndexes(final String indexWildCard) throws IOException {
    return client
        .cat()
        .indices()
        .thenAccept(
            response ->
                response.valueBody().stream()
                    .map(IndicesRecord::index)
                    .filter(index -> index.matches(indexWildCard))
                    .toList());

    return indices;
  }
```

Reply:
Good point all these are returning completable futures!